plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServiceViewModel.java

```diff
@@ -85,7 +97,44 @@ private void updateServiceItems(List<ObservableTreeItem> newServiceItems) {
 	}
 
 	protected void updateService(IService service, List<ObservableTreeItem> serviceItems) {
-		firePropertyChange(PROPERTY_SERVICE, null, this.service = getServiceOrDefault(service, serviceItems));
+		IService newService = getServiceOrDefault(service, serviceItems);
+		if(newService != this.service) {
+			boolean needUpdateRoutes = newService == null || this.service == null || newService.getProject() != this.service.getProject();
+			firePropertyChange(PROPERTY_SERVICE, null, this.service = newService);
+			if(needUpdateRoutes) {
+				updateRoutes();
+			}
+		}
+	}
+
+	protected void updateRoutes() {
+		if(this.service == null && !routes.isEmpty()) {
+			List<IRoute> oldRoutes = new ArrayList<>(this.routes);
+			this.routes.clear();
+			firePropertyChange(PROPERTY_ROUTES, oldRoutes, this.routes);
```

Comment:
we're updating the selected route while the method claims to update the collection of available routes. I'd rather move this down to #updateRoute or get both methods unified for added code clarity (ex. I'd try to fix an issue with the selected route, would dig into updateRoute() and then unexpectely find out that the route is also set in #updateRoutes. Not ideal imho.